pkg/controller/rollout/rollout_bluegreen.go

```diff
@@ -0,0 +1,487 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	utilpointer "k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type blueGreenReleaseManager struct {
+	client.Client
+	trafficRoutingManager *trafficrouting.Manager
+	recorder              record.EventRecorder
+}
+
+func (m *blueGreenReleaseManager) runCanary(c *RolloutContext) error {
+	blueGreenStatus := c.NewStatus.BlueGreenStatus
+	if br, err := fetchBatchRelease(m.Client, c.Rollout.Namespace, c.Rollout.Name); err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("rollout(%s/%s) fetch batchRelease failed: %s", c.Rollout.Namespace, c.Rollout.Name, err.Error())
+		return err
+	} else if err == nil {
+		// This line will do something important:
+		// - sync status from br to Rollout: to better observability;
+		// - sync rollout-id from Rollout to br: to make BatchRelease
+		//   relabels pods in the scene where only rollout-id is changed.
+		if err = m.syncBatchRelease(br, blueGreenStatus); err != nil {
+			klog.Errorf("rollout(%s/%s) sync batchRelease failed: %s", c.Rollout.Namespace, c.Rollout.Name, err.Error())
+			return err
+		}
+	}
+	if blueGreenStatus.PodTemplateHash == "" {
+		blueGreenStatus.PodTemplateHash = c.Workload.PodTemplateHash
+	}
+
+	if m.doCanaryJump(c) {
+		klog.Infof("rollout(%s/%s) canary step jumped", c.Rollout.Namespace, c.Rollout.Name)
+		return nil
+	}
+	// When the first batch is trafficRouting rolling and the next steps are rolling release,
+	// We need to clean up the canary-related resources first and then rollout the rest of the batch.
+	currentStep := c.Rollout.Spec.Strategy.BlueGreen.Steps[blueGreenStatus.CurrentStepIndex-1]
+	if currentStep.Traffic == nil && len(currentStep.Matches) == 0 {
+		tr := newTrafficRoutingContext(c)
+		done, err := m.trafficRoutingManager.FinalisingTrafficRouting(tr)
+		blueGreenStatus.LastUpdateTime = tr.LastUpdateTime
+		if err != nil {
+			return err
+		} else if !done {
+			klog.Infof("rollout(%s/%s) cleaning up canary-related resources", c.Rollout.Namespace, c.Rollout.Name)
+			expectedTime := time.Now().Add(tr.RecheckDuration)
+			c.RecheckTime = &expectedTime
+			return nil
+		}
+	}
+
+	switch blueGreenStatus.CurrentStepState {
+	// before CanaryStepStateUpgrade, handle some special cases, to prevent traffic loss
+	case v1beta1.CanaryStepStateInit:
+		klog.Infof("rollout(%s/%s) run bluegreen strategy, and state(%s)", c.Rollout.Namespace, c.Rollout.Name, v1beta1.CanaryStepStateInit)
+		tr := newTrafficRoutingContext(c)
+		if currentStep.Traffic == nil && len(currentStep.Matches) == 0 {
+			blueGreenStatus.CurrentStepState = v1beta1.CanaryStepStateUpgrade
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				blueGreenStatus.CurrentStepIndex, v1beta1.CanaryStepStateInit, blueGreenStatus.CurrentStepState)
+			return nil
```

Comment:
consider using a ’labeled’ fallthrough to avoid unnecessary requeue , e.g 
```
if currentStep.Traffic == nil && len(currentStep.Matches) == 0 {
   ...
   goto nextCase
}
....
nextCase:
  fallthrough
case: v1beta1.CanaryStepStateUpgrade:

```

Reply:
ok